sns_ik_lib/src/sns_position_ik.cpp

```diff
@@ -145,30 +162,30 @@ int SNSPositionIK::CartToJnt(const KDL::JntArray& joint_seed,
 
     }
 
-    //sf = m_ikVelSolver->getJointVelocity(&qDot, sot, q_i.data);
     m_ikVelSolver->getJointVelocity(&qDot, sot, q_i.data);
 
-    q_i.data += m_dt * qDot;
-
-    for (int j = 0; j < jl_low.rows(); ++j) {
-      q_i.data[j] = std::max(std::min(q_i.data[j], jl_high[j]), jl_low[j]);
-    }
-
-    //std::cout << "    q: " << q_i.data.transpose() << std::endl;
-    //std::cout << "    cart vel: " << sot[0].desired.transpose() << std::endl;
-    //std::cout << "    qDot: " << qDot.transpose() << std::endl;
-
-    if (qDot.norm() < 1e-5) {  // TODO: config param
+    if (qDot.norm() < 1e-6) {  // TODO: config param
       //std::cout << "ERROR: Solution stuck, iter: "<<ii<<", error: " << lineErr << " m, " << rotErr << " rad" << std::endl;
       return -2;
     }
 
-//    if (sf < stepScale) {
-//      stepScale = std::max(0.8*stepScale, 0.01);
-//      //std::cout << "New step scale (" << ii << "): " << stepScale << std::endl;
-//    } else if (stepScale < 1.0) {
-//      stepScale = std::min(1.1*stepScale, 1.0);
-//    }
+    q_i.data += m_dt * qDot;
+
+    // Apply a decaying barrier function
+    // u = upper limit;  l = lower limit
+    // theta(x) = -log(u - x) - log(-l + x)
+    // -alpha * dTheta(x)/dx = alpha*(1/(x-l) + 1/(x-u))
+    if (m_useBarrierFunction && (lineErr > 0.5 || rotErr > 0.5) ) {
+      for (int j = 0; j < jl_low.rows(); ++j) {
+        q_i.data[j] = std::max(std::min(q_i.data[j], jl_high[j] - 1e-7), jl_low[j] + 1e-7);
+        q_i.data[j] += barrierAlpha * (1/(q_i.data[j] - jl_low[j]) + 1/(q_i.data[j] - jl_high[j]));
+      }
+      barrierAlpha *= m_barrierDecay;
```

Comment:
If for some reason m_useBarrierFunction but m_barrierDecay was also set to zero, then nothing will be added to q_i and this won't have any affect.

m_barrierDecay needs to be in the range (0.0, 1.0], but this should be checked in the setter and not here.